### PR TITLE
Prepare interpreter for function calling

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21515 Add WeakIValue, use in tracer.
* #21514 clean up the TracingState API
* #21513 Collapse tracing_state.h into tracer.h
* #21325 Interpreter support for CallFunction/CallMethod
* #21273 Expose ExecutionPlan in prep for function calls
* #21272 Add flag to temporarily enable first class modules
* #21240 Reduce number of stack manipulation instructions in interpreter.
* **#21185 Prepare interpreter for function calling**

Summary: In order to support calling functions without inlining,
the interpreter will need to push another function 'frame' somewhere
when we invoke the call operator. This behavior is not easily expressible
in the current interpreter since ops can only push/pop to the stack.

Furthermore, we have accumulated several other warts because of the design:

* every Operation needs to return an int because a few ops branch
* aten::wait has to communicate with the core interpreter loop via exceptions.
* The way we encode Loop is a complicated desugaring because implementing
  it as an operator that can only push/pop the stack is hard.

To enable an easy implementation of the call instruction, and to fix the
above warts this patch changes the format of the interpreter instructions
to make it easier to encode arbitrary functionality.

Notes:
* Instructions are now 64-bit words with an opcode, and two arguments
* The var-arg register push/pop/move parts of the instructions are gone.
  Explicit instructions LOAD/STORE/MOVE now encode this logic.
* LOOP/WAIT are directly implemented as operators. We no longer desugar.
* We no longer need to lower Param and Return nodes, avoiding one
  preprocessing pass.
* Debug info is stored off to the side since now multiple Instructions
  may correspond to one node in the graph.

Future:
* This patch should not be merged on its own. A followup will add a
  pass that optimizes the register push/pop logic to cut down the
  number of raw instructions. The pretty printer already does something
  similar to inline instructions.
* The follow up patch will require performance testing before merging
  to make sure this does not regress interpreter overhead.

Test Plan: test_jit.py

Differential Revision: [D15572818](https://our.internmc.facebook.com/intern/diff/D15572818)

### CODE REVIEW

**Review by suo — APPROVED** [review]
Looks great! I really like how much easier the code is to follow now

File: torch/csrc/jit/interpreter.cpp [hunk 1]
```diff
@@ -290,448 +196,499 @@ std::unordered_map<Node*, std::vector<uint8_t>> findLastUses(Graph& g) {
...
+// R - index into register table
+// I - literal integer
+// C - index into constant table
+// P - jump offset relative to beginning of current instruction
+
+#define FORALL_OPCODES(_)                                                   \
+  _(OP, "O") /* invoke operator X */                                        \
+  _(LOAD, "R") /* push a value from a register X */                         \
+  _(MOVE, "R") /* push a value from register X, clearing the register */    \
+  _(STOREN, "RI") /* store N values to registers [X, X+N) */                \
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: Presumably STORE is slightly faster for storing a single element and almost everything thing returns one argument.

File: torch/csrc/jit/interpreter.cpp [hunk 2]
```diff
@@ -290,448 +196,499 @@ std::unordered_map<Node*, std::vector<uint8_t>> findLastUses(Graph& g) {
...
+  _(STORE, "R") /* store 1 value to registers X */                          \
+  _(DROP, "") /* drop 1 value from the top of the stack */                  \
+  _(DROPR, "R") /* clear register X */                                      \
+  _(LOADC, "C") /* push the constant X */                                   \
+  _(JF, "P") /* pop the top of the stack, if false, branch to P */          \
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: I can change it to JMP_IF_FALSE.

File: torch/csrc/jit/interpreter.cpp [hunk 3]
```diff
@@ -290,448 +196,499 @@ std::unordered_map<Node*, std::vector<uint8_t>> findLastUses(Graph& g) {
...
+
+  // same length as instructions.
+  // what node in the graph cause this
+  // instruction to be emitted?
+  std::vector<Node*> instructions_source_;
+
+  std::vector<IValue> constant_table_;
+  std::vector<Operation> operator_table_;
+  int register_size_ = 0;
+  size_t n_outputs;
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: I think the general idea is to not rely on having the graph_ except for debugging, making it clearer what the interpreter needs from the graph.

File: torch/csrc/jit/interpreter.cpp [hunk 4]
```diff
@@ -290,448 +196,499 @@ std::unordered_map<Node*, std::vector<uint8_t>> findLastUses(Graph& g) {
...
+            if (num_outputs == 1) {
+              future->markCompleted(stack.back());
+            } else {
+              future->markCompleted(
+                  Tuple::create(jit::last(stack, num_outputs).vec()));
+            }
+          }
+          return false;
+        case WAIT:
+          auto future = stack.back().toFuture();
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: I think the followup PR already does this.

**Pull request closed without merging**